### PR TITLE
zq: Better error messages for single argument commands

### LINKED ISSUE #5046: Zed syntax error causes in confusing error message about file input
## tl;dr

A community user was confused by this error message:

```
$ zq -version
Version: v1.14.0-2-g4dc62369

$ zq -Z 'fork(=> file "./csv_1.csv" | yield email | collect(this), => file "./csv_2.csv" | yield this["Recommended Email"])'
zq: no such file: fork(=> file "./csv_1.csv" | yield email | collect(this), => file "./csv_2.csv" | yield this["Recommended Email"])
```

## Details

Repro is with Zed commit 4dc6236. The issue was initially reported in a [community Slack thread](https://brimdata.slack.com/archives/CTT24051C/p1708106640649629).

Many `zq` command lines are unambiguous when the first argument is the Zed program text and the remaining arguments are file inputs. The example above presents a challenge because there's just a single argument, so `zq` needs to use some kind of heuristic to determine how to treat that single argument. It could be a filename (e.g., a simple command line like `zq MyFile.zng` may often be used to view a file's contents as ZSON) or, as was clearly the user's intent here, it could be a Zed program that specifies the inputs from inside using the `file` operator. In this case there was a syntax error in the Zed program (the `,` before the `=>`) so the heuristic fell back to treating the single input as a filename, and the current error message about this understandably confused the user.

We discussed this one as a group and there was some thought that the heuristic could maybe stand to be revisited and/or a new flag could be introduced similar to `jq -n`. However, as a short term improvement there was some consensus that we might get by just improving the error message here to be explicit about what the heuristic attempted to make it more obvious to the user in a situation like this that it's actually a Zed syntax error. Most of the error messages from the Zed CLI tooling are fairly terse, but this situation is tricky enough that it may be one where some verbosity is justified.

### PR DESCRIPTION
**Pull request by mattnibs**
Fixes #5046

### CODE REVIEW

**Comment by philrz** [comment]
I just tested this branch out at commit f361f112 using the repro example from #5046.

```
$ zq -version
Version: v1.14.0-7-gf361f112

$ zq -Z 'fork(=> file "./csv_1.csv" | yield email | collect(this), => file "./csv_2.csv" | yield this["Recommended Email"])'
zq: could not invoke zq with a single argument because:
 - the argument did not parse as a valid Zed query
 - a file could not be found with the name "fork(=> file \"./csv_..."
```

This seems like a great improvement to me so I'm functionally 👍 on this.

As @mattnibs and I have discussed offline, the user would surely appreciate being able to see the parse error in a situation like this. Two thoughts come to mind:

1. We could present it as part of this two-part message. If we do, I might suggest swapping the order since I expect the parse error will take up multiple lines.
2. We could implement the `-query` flag as mention in the docs and #5052 and make some reference to it in this error message (e.g., "consider `-query`").

**Comment by philrz** [comment]
I just re-tested the branch now at commit 2699090 once again using that example from #5046. It looks great with the parse error!

```
$ zq -version
Version: v1.14.0-8-g2699090d

$ zq -Z 'fork(=> file "./csv_1.csv" | yield email | collect(this), => file "./csv_2.csv" | yield this["Recommended Email"])'
zq: could not invoke zq with a single argument because:
 - a file could not be found with the name "fork(=> file \"./csv_..."
 - the argument could not be compiled as a valid Zed query due to parse error (column 58):
   fork(=> file "./csv_1.csv" | yield email | collect(this), => file "./csv_2.csv" | yield this["Recommended Email"])
                                                         === ^ ===
```

I then thought to see how it would look if I had some multi-line Zed.

```
$ zq -Z 'fork (
>   => file "./csv_1.csv" | yield email | collect(this),
>   => file "./csv_2.csv" | yield this["Recommended Email"]
> )'
zq: could not invoke zq with a single argument because:
 - a file could not be found with the name "fork (\n  => file \"./..."
 - the argument could not be compiled as a valid Zed query due to parse error (line 2, column 2):
     => file "./csv_2.csv" | yield this["Recommended Email"]
   = ^ ===
```

That's great since it mentions the line & column but shows just the line that had the error. Very nice! 👍 

(I guess it's a partial bummer that the parser doesn't point directly at the `,` as being the problem, but that doesn't appear to be this PR's fault.)

File: cmd/zq/ztests/single-arg-error.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  ! zq 'file sample.zson | count('
+
+outputs:
+  - name: stderr
+    data: |
+      zq: could not invoke zq with a single argument because:
+       - a file could not be found with the name "file sample.zson | c..."
+       - the argument could not be compiled as a valid Zed query due to parse error (column 25):
+         file sample.zson | count(
+                              === ^ ===
```
Comment (nwt) [inline cmd/zq/ztests/single-arg-error.yaml:11]: This is pretty wordy. How about something more concise?
```
zq: not a file or valid query: "from ( file a ) | co..."
  error parsing Zed at column 25:
  from ( file a ) | count(
                      === ^ ===
```

As a bonus, that would make for a much simpler `singleArgError`.
```go
func singleArgError(src string, err error) error {
	if len(src) > 20 {
		src = src[:20] + "..."
	}
	msg := fmt.Sprintf("not a file or valid query: %q", src)
	if errors.As(err, new(*parser.Error)) {
		msg += "\n" + err.Error()
	}
	msg = strings.ReplaceAll(msg, "\n", "\n   ")
	return errors.New(msg)
}
```

Comment (nwt) [reply cmd/zq/ztests/single-arg-error.yaml:11]: (To be clear, I'm OK with the existing message if you guys prefer it.)

Comment (philrz) [reply cmd/zq/ztests/single-arg-error.yaml:11]: Thanks @nwt. Indeed, I was in the middle of writing a reply when that last comment of yours came in, so I'll paste that below for completeness, but basically, yeah, I happen to prefer what's already there. 😄

I know I can tend toward wordiness in my own work and so I try to keep that in check, but I think this is one place where verbosity is maybe justified. For instance, in the #5046 case the user is almost certainly thinking entirely in terms of their argument being a Zed query so I'm concerned something concise like "not a file" might still be confusing. "Not a file? What? The file refs are in my program. I don't get it." This particular error is triggered rarely enough and has confused users enough in is previous form in enough that I feel pretty strongly that the benefits of super clear communication outweigh any assumed benefits from being concise.

Come to think of it, the user that last hit #5046 is responsive on Slack today, so I'll see if they're game to weigh in to the discussion, though if we already have consensus and want to get on with it and merge I'm fine with that too.

Comment (nwt) [reply cmd/zq/ztests/single-arg-error.yaml:11]: @mattnibs: Any preference here?

**Review by nwt — APPROVED** [review]

File: cmd/zq/ztests/single-arg-error.yaml (hunk 1 above)
Comment (iloveitaly) [reply cmd/zq/ztests/single-arg-error.yaml:11]: fwiw, as someone who doesn't use zed every day, but reaches for it probably once a week when manipulating data I'd prefer the more verbose message. It's hard to understand exactly what might be wrong, especially given the unfamiliarity of the zq syntax compared to other tools.

**Pull request merged**